lib/ASRouterTriggerListeners.jsm

```diff
@@ -0,0 +1,142 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+ChromeUtils.import("resource://gre/modules/Services.jsm");
+
+ChromeUtils.defineModuleGetter(this, "PrivateBrowsingUtils",
+  "resource://gre/modules/PrivateBrowsingUtils.jsm");
+
+/**
+ * A Map from trigger IDs to singleton trigger listeners. Each listener must
+ * have idempotent `init` and `uninit` methods.
+ */
+const TriggerListeners = new Map([
+  ["openURL", {
+    /**
+     * Attach listeners to every browser window to detect location changes, and
+     * notify the trigger handler whenever we navigate to a URL with a hostname
+     * we're looking for.
+     */
+    _initialized: false,
+    _triggerHandler: null,
+    _hosts: new Set(),
+
+    init(triggerHandler, hosts = new Set()) {
+      if (!this._initialized) {
+        this.onLocationChange = this.onLocationChange.bind(this);
+
+        // Listen for new windows being opened
+        Services.ww.registerNotification(this);
+
+        // Add listeners to all existing browser windows
+        const winEnum = Services.wm.getEnumerator("navigator:browser");
+        while (winEnum.hasMoreElements()) {
+          let win = winEnum.getNext();
+          if (win.closed || PrivateBrowsingUtils.isWindowPrivate(win)) {
+            continue;
+          }
+          win.gBrowser.addTabsProgressListener(this);
+        }
+
+        this._initialized = true;
+      }
+      this._triggerHandler = triggerHandler;
+      this._hosts = new Set(hosts); // Clone the hosts to avoid unexpected behaviour
+    },
+
+    uninit() {
+      if (this._initialized) {
+        Services.ww.unregisterNotification(this);
+
+        const winEnum = Services.wm.getEnumerator("navigator:browser");
+        while (winEnum.hasMoreElements()) {
+          let win = winEnum.getNext();
+          if (win.closed || PrivateBrowsingUtils.isWindowPrivate(win)) {
+            continue;
+          }
+
+          win.gBrowser.removeTabsProgressListener(this);
+        }
+
+        this._initialized = false;
+        this._triggerHandler = null;
+        this._hosts.clear();
+      }
+    },
+
+    onLocationChange(aBrowser, aWebProgress, aRequest, aLocationURI, aFlags) {
+      const location = aLocationURI ? aLocationURI.spec : "";
+      if (aWebProgress.isTopLevel) {
+        const host = (new URL(location)).hostname;
+        if (this._hosts.has(host)) {
+          this._triggerHandler(aBrowser.messageManager, {id: "openURL", param: host});
+        }
+      }
+    },
+
+    observe(subject, topic, data) {
+      let win;
+      let onLoad;
+
+      switch (topic) {
+        case "domwindowopened":
+          win = subject.QueryInterface(Ci.nsIDOMWindow);
```

Comment:
Actually, as of recently, the subject is the window! no need to QueryInterface anymore